### PR TITLE
Use a non member mutex lock for serverConfig access.

### PR DESCRIPTION
**Pull request by hackintoshrao**
Use a non member mutex lock for serverConfig access.
## Description

 - This is to ensure that the any new config references made to the
      serverConfig is also backed by a mutex lock.
 - Otherwise any new config assigment will also replace the member mutex
      which is currently used for safe access.
## Motivation and Context

- Useful to also guarantee that tests don't influence each other by new assignment of config.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by hackintoshrao** [comment]
Part - 1 necessary changes for https://github.com/minio/minio/issues/3198 .

File: cmd/config-v10.go [hunk 1]
```diff
@@ -23,8 +23,11 @@ import (
 	"github.com/minio/minio/pkg/quick"
 )
 
+// Read Write mute for safe access to ServerConfig.
```
Comment (donatello) [inline cmd/config-v10.go]: `mutex` instead of `mute`.

**Review by donatello — APPROVED** [review]
LGTM - except for typo in comment.

File: cmd/config-v10.go [hunk 2]
```diff
@@ -23,8 +23,11 @@ import (
...
+var severConfigMu sync.RWMutex
```
Comment (krisis) [inline cmd/config-v10.go]: typo: s/severConfigMu/serverConfigMu

File: cmd/config-v10.go [hunk 3]
```diff
@@ -112,168 +117,191 @@ var serverConfig *serverConfigV10
 
 // GetVersion get current config version.
 func (s serverConfigV10) GetVersion() string {
-	s.rwMutex.RLock()
-	defer s.rwMutex.RUnlock()
+	severConfigMu.Lock()
```
Comment (krisis) [inline cmd/config-v10.go]: This should be a RLock()/RUnlock() here and in other places where members of serverConfigV10 objects are being read.

**Review by krisis — CHANGES_REQUESTED** [review]

File: cmd/config-v10.go [hunk 4]
```diff
@@ -37,13 +40,11 @@ type serverConfigV10 struct {
...
 	// Notification queue configuration.
 	Notify notifier `json:"notify"`
-
-	// Read Write mutex.
-	rwMutex *sync.RWMutex
 }
 
 // initConfig - initialize server config and indicate if we are creating a new file or we are just loading
 func initConfig() (bool, error) {
+
```
Comment (harshavardhana) [inline cmd/config-v10.go]: Unnecessary new line.

File: cmd/config-v10.go [hunk 5]
```diff
@@ -68,17 +69,18 @@ func initConfig() (bool, error) {
...
 		// Create config path.
 		err := createConfigPath()
 		if err != nil {
 			return false, err
 		}
-
+		// hold the mutex lock before a new config is assigned.
+		serverConfigMu.Lock()
 		// Save the new config globally.
 		serverConfig = srvCfg
```
Comment (harshavardhana) [inline cmd/config-v10.go]: Move this up a bit assign to the variable like this

```
serverConfigMu.Lock()
serverConfig = srvCfg
serverConfigMu.Unlock()
```

File: cmd/config-v10.go (hunk 4 above)
Comment (hackintoshrao) [reply cmd/config-v10.go]: Removing.

File: cmd/config-v10.go (hunk 5 above)
Comment (hackintoshrao) [reply cmd/config-v10.go]: Does it mean that the assignment with the mutex locks should be in 3 consecutive lines?

**Review by krisis — APPROVED** [review]

File: cmd/config-v10.go (hunk 5 above)
Comment (harshavardhana) [reply cmd/config-v10.go]: Yes that's why it is readable.

Comment (hackintoshrao) [reply cmd/config-v10.go]: done.

**Pull request merged**